beacon-chain/rpc/eventsv1/events.go

```diff
@@ -0,0 +1,204 @@
+}
+
+// StreamEvents allows requesting all events from a set of topics defined in the eth2.0-apis standard.
+// The topics supported include block events, attestations, chain reorgs, voluntary exits,
+// chain finality, and more.
+func (s *Server) StreamEvents(
+	req *ethpb.StreamEventsRequest, stream ethpb.Events_StreamEventsServer,
+) error {
+	if req == nil || len(req.Topics) == 0 {
+		return status.Error(codes.InvalidArgument, "no topics specified to subscribe to")
+	}
+	// Check if the topics in the request are valid.
+	requestedTopics := make(map[string]bool)
+	for _, topic := range req.Topics {
+		if _, ok := casesHandled[topic]; !ok {
+			return status.Errorf(codes.InvalidArgument, "topic %s not allowed for event subscriptions", topic)
+		}
+		requestedTopics[topic] = true
+	}
+
+	// Subscribe to event feeds from information received in the beacon node runtime.
+	blockChan := make(chan *feed.Event, 1)
+	blockSub := s.BlockNotifier.BlockFeed().Subscribe(blockChan)
+
+	opsChan := make(chan *feed.Event, 1)
+	opsSub := s.OperationNotifier.OperationFeed().Subscribe(opsChan)
+
+	stateChan := make(chan *feed.Event, 1)
+	stateSub := s.StateNotifier.StateFeed().Subscribe(stateChan)
+
+	defer blockSub.Unsubscribe()
+	defer opsSub.Unsubscribe()
+	defer stateSub.Unsubscribe()
+
+	// Handle each event received and context cancelation.
+	for {
+		select {
+		case event := <-blockChan:
+			if err := s.handleBlockEvents(stream, requestedTopics, event); err != nil {
+				return status.Errorf(codes.Internal, "could not handle block event: %v", err)
+			}
+		case event := <-opsChan:
+			if err := s.handleBlockOperationEvents(stream, requestedTopics, event); err != nil {
+				return status.Errorf(codes.Internal, "could not handle block operations event: %v", err)
+			}
+		case event := <-stateChan:
+			if err := s.handleStateEvents(stream, requestedTopics, event); err != nil {
+				return status.Errorf(codes.Internal, "could not handle state event: %v", err)
+			}
+		case <-s.Ctx.Done():
+			return status.Errorf(codes.Canceled, "context canceled")
+		case <-stream.Context().Done():
+			return status.Errorf(codes.Canceled, "context canceled")
+		}
+	}
+}
+
+func (s *Server) handleBlockEvents(
+	stream ethpb.Events_StreamEventsServer, requestedTopics map[string]bool, event *feed.Event,
+) error {
+	switch event.Type {
+	case blockfeed.ReceivedBlock:
+		if _, ok := requestedTopics[blockTopic]; !ok {
+			return nil
+		}
+		blkData, ok := event.Data.(*blockfeed.ReceivedBlockData)
+		if !ok {
+			return nil
+		}
+		v1Data, err := migration.BlockIfaceToV1BlockHeader(blkData.SignedBlock)
+		if err != nil {
+			return err
+		}
+		item, err := v1Data.HashTreeRoot()
+		if err != nil {
+			return status.Errorf(codes.Internal, "could not hash tree root block %v", err)
+		}
+		eventBlock := &ethpb.EventBlock{
+			Slot:  v1Data.Message.Slot,
+			Block: item[:],
+		}
+		return s.streamData(stream, blockTopic, eventBlock)
+	default:
+		return nil
```

Comment:
Perhaps default should return error for better visibility. Same for the below defaults